File: src/_01_CreationalPattern/_01_04_Builder/java/after/TourPlan.java
package _01_CreationalPattern._01_04_Builder.java.after;

import java.time.LocalDate;
import java.util.List;

public class TourPlan {
  
  private String title;
  private int nights;
  private int days;
  private LocalDate startDate;
  private String whereToStay;
  private List<DetailPlan> plans;

  public TourPlan() {}

  public TourPlan(String title, int nights, int days, LocalDate startDate, String whereToStay, List<DetailPlan> plans) {
    this.title = title;
    this.nights = nights;
    this.days = days;
    this.startDate = startDate;
    this.whereToStay = whereToStay;
    this.plans = plans;
  }

  @Override
  public String toString() {
    return "TourPlan {"
      + " title=\"" + title + "\""
      + ", nights=" + nights
      + ", days=" + days
      + ", startDate=" + startDate
      + ", whereToStay=\"" + whereToStay + "\""
      + ", plans=" + plans
      + " }";
  }

  // title
  public String getTitle() {
    return title;
  }
  public void setTitle(String title) {
    this.title = title;
  }

  // nights
  public int getNights() {
    return nights;
  }
  public void setNights(int nights) {
    this.nights = nights;
  }

  // days
  public int getDays() {
    return days;
  }
  public void setDays(int days) {
    this.days = days;
  }

  // startDate
  public LocalDate getStartDate() {
    return startDate;
  }
  public void setStartDate(LocalDate startDate) {
    this.startDate = startDate;
  }

  // whereToStay
  public String getWhereToStay() {
    return whereToStay;
  }
  public void setWhereToStay(String whereToStay) {
    this.whereToStay = whereToStay;
  }

  // plans
  public List<DetailPlan> getPlans() {
    return plans;
  }
  public void setPlans(List<DetailPlan> plans) {
    this.plans = plans;
  }

  public void addPlan(int day, String plan) {
    plans.add(new DetailPlan(day, plan));
  }

}

File: src/_01_CreationalPattern/_01_04_Builder/TourPlan.js
import DetailPlan from "./DetailPlan.js";

export default class TourPlan {
  /** @type { string } */
  #title;

  /** @type { number } */
  #nights;

  /** @type { number } */
  #days;

  /** @type { Date } */
  #startDate;

  /** @type { string } */
  #whereToStay;

  /** @type { DetailPlan[] } */
  #plans;

  /** @returns { string } */
  printInfo() {
    return `
      TourPlan: {
        title: ${this.#title},
        nights: ${this.#nights},
        days: ${this.#days},
        startDate: ${this.#startDate},
        whereToStay: ${this.#whereToStay},
        plans: ${this.#plans?.map(plan => plan.printInfo())},
      }
    `;
  }

  /** @returns { string } */
  getTitle() {
    return this.#title;
  }
  /** @param { string } title */
  setTitle(title) {
    this.#title = title;
  }

  /** @returns { number } nights */
  getNights() {
    return this.#nights;
  }
  /** @param { number } nights */
  setNights(nights) {
    this.#nights = nights;
  }

  /** @returns { number } */
  getDays() {
    return this.#days;
  }
  /** @param { number } days */
  setDays(days) {
    this.#days = days;
  }

  /** @returns { Date } */
  getStartDate() {
    return this.#startDate;
  }
  /** @param { Date } startDate */
  setStartDate(startDate) {
    this.#startDate = startDate;
  }

  /** @returns { string } */
  getWhereToStay() {
    return this.#whereToStay;
  }
  /** @param { string } _whereToStay */
  setWhereToStay(_whereToStay) {
    this.#whereToStay = _whereToStay;
  }

  /** @returns { DetailPlan[] } */
  getPlans() {
    return this.#plans;
  }
  /** @param { DetailPlan[] } plans */
  setPlans(plans) {
    this.#plans = plans;
  }

  /**
   * @param { number } day
   * @param { string } plan
   */
  addPlan(day, plan) {
    this.#plans ?? (this.#plans = []);
    this.#plans.push(new DetailPlan(day, plan));
  }
}